scripts/zkai_attention_derived_d128_block_statement_chain_gate.py

```
@@ -0,0 +1,921 @@
+    if not stat_module.S_ISDIR(parent_stat.st_mode):
+        raise AttentionDerivedD128BlockStatementChainError(f"output parent must be a directory: {resolved.parent}")
+    dir_fd = os.open(
+        resolved.parent,
+        os.O_RDONLY | getattr(os, "O_DIRECTORY", 0) | getattr(os, "O_NOFOLLOW", 0),
+    )
+    tmp_name = f".{resolved.name}.{os.getpid()}.{secrets.token_hex(8)}.tmp"
+    tmp_created = False
+    try:
+        open_parent_stat = os.fstat(dir_fd)
+        if (open_parent_stat.st_dev, open_parent_stat.st_ino) != (parent_stat.st_dev, parent_stat.st_ino):
+            raise AttentionDerivedD128BlockStatementChainError(f"output parent changed while opening: {resolved.parent}")
```

Ruff (0.15.12)

[warning] Abstract `raise` to an inner function

(TRY301)